File: app/src/androidTest/java/it/cosenonjaviste/androidtest/base/FragmentRule.java
package it.cosenonjaviste.androidtest.base;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.test.InstrumentationRegistry;
import android.support.test.rule.ActivityTestRule;
import android.support.v4.app.Fragment;

import it.cosenonjaviste.core.base.ViewModel;
import it.cosenonjaviste.ui.utils.SingleFragmentActivity;

public class FragmentRule extends ActivityTestRule<SingleFragmentActivity> {
    private Class<? extends Fragment> fragmentClass;

    public FragmentRule(Class<? extends Fragment> fragmentClass) {
        super(SingleFragmentActivity.class, false, false);
        this.fragmentClass = fragmentClass;
    }

    public SingleFragmentActivity launchFragment() {
        return launchFragment(null);
    }

    public void launchFragment(Parcelable model) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(ViewModel.MODEL, model);
        launchFragment(bundle);
    }

    public SingleFragmentActivity launchFragment(Bundle b) {
        Intent intent = SingleFragmentActivity.populateIntent(new Intent(), fragmentClass);
        if (b != null) {
            intent.putExtras(b);
        }
        return launchActivity(intent);
    }

    public <T extends Context> T getApplication() {
        return (T) InstrumentationRegistry.getTargetContext().getApplicationContext();
    }
}

File: app/src/androidTest/java/it/cosenonjaviste/androidtest/base/FragmentRule.kt
package it.cosenonjaviste.androidtest.base

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.support.test.rule.ActivityTestRule
import android.support.v4.app.Fragment

import it.cosenonjaviste.mv2m.ViewModel
import it.cosenonjaviste.ui.utils.SingleFragmentActivity

class FragmentRule(private val fragmentClass: Class<out Fragment>) : ActivityTestRule<SingleFragmentActivity>(SingleFragmentActivity::class.java, false, false) {

    fun launchFragment(model: Parcelable) {
        val bundle = Bundle()
        bundle.putParcelable(ViewModel.MODEL, model)
        launchFragment(bundle as Bundle?)
    }

    @JvmOverloads fun launchFragment(b: Bundle? = null): SingleFragmentActivity {
        val intent = SingleFragmentActivity.populateIntent(Intent(), fragmentClass)
        if (b != null) {
            intent.putExtras(b)
        }
        return launchActivity(intent)
    }
}
